### PR TITLE
Adding tests for mass calculator

### LINKED ISSUE #53: Add test for MassCalculator.tsx
https://github.com/OpenSourceRaidGuild/atom-pwa/projects/1#card-55812284

### PR DESCRIPTION
**Pull request by MohitPopli**
Screen Shot 2021-02-28 at 2 50 02 PM

Lines left are edge cases and would like to discuss if anyone have idea.

closes #53

### CODE REVIEW

File: src/components/mass-calculator/MassCalculator.test.tsx [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+];
+
+beforeEach(() => {
+  jest.clearAllMocks();
+  jest
+    .spyOn(ElementManager, "getElement")
+    .mockImplementation((atomic) =>
+      elementsArr.find((element) => element.atomic === atomic)
+    );
+});
```
Comment (JacobMGEvans) [inline src/components/mass-calculator/MassCalculator.test.tsx]: @mpeyper this is concerning to me, its a code implementation detail AND relies on the class/function having getter methods

Comment (JacobMGEvans) [reply src/components/mass-calculator/MassCalculator.test.tsx]: IF there isn't a better way I understand using this. 

@nobrayner  @Aprillion  @codyarose Thoughts?

Comment (MohitPopli) [reply src/components/mass-calculator/MassCalculator.test.tsx]: I'll  be glad to have some suggestions over this. As `ElementManager` is exposed as instance object with getter methods and doing my research I ended up spying over it.
Also that `hammer.js` gave me headache as in how to inform jest about Hammer. But finally that one's resolved.

Comment (JacobMGEvans) [reply src/components/mass-calculator/MassCalculator.test.tsx]: > I'll glad to have some suggestions over this. As `ElementManager` is exposed as instance object with getter methods and doing my research I ended up spying over it.
> Also that `hammer.js` gave me headache as in how to inform jest about Hammer. But finally that one's resolved.

These tests are also for refactoring to Hooks, so making tests around Class-based models is not ideal either.

Comment (JacobMGEvans) [reply src/components/mass-calculator/MassCalculator.test.tsx]: from the looks of the code, it gets the Element in `massCalculatorElement`
```js
 private massCalculatorElement(massCalculatorElement: IMassCalculatorElement) {
    const { atomic, quantity } = massCalculatorElement;
// Here is what you are Spying on and changing behavior
    const element = ElementManager.getElement(atomic); 
```
`atomic` is part of the components initial state, which you can more reliably mock 
```js
interface IMassCalculatorElement {
  atomic: number; // Used for the ElementManager getter
  quantity: number;
}

type Props = RouteComponentProps<any> & React.Props<any>;

interface IMassCalculatorState {
  elements: IMassCalculatorElement[]; // which is part of the initial state
  addElementModalOpen: boolean;
  modifyElementModal: {
    open: boolean;
    elementIndex: number;
  };
}
```

Comment (JacobMGEvans) [reply src/components/mass-calculator/MassCalculator.test.tsx]: Here is the INITIAL location utilizing Atomic from state which creates the `massCalculatorElement` used later for other `ElementManager` getter
```js 
// elements is destructured from state
  const massCalculatorElement = elements[index]; 

    if (!massCalculatorElement) {
      return null;
    }

    const element = ElementManager.getElement(massCalculatorElement.atomic); 
```

**Review by JacobMGEvans — CHANGES_REQUESTED** [review]
*See comments in thread https://github.com/OpenSourceRaidGuild/atom-pwa/pull/54#discussion_r584175489

File: src/components/mass-calculator/MassCalculator.test.tsx (hunk 1 above)
Comment (grady-lad) [reply src/components/mass-calculator/MassCalculator.test.tsx]: It looks like `ElementManager` is a singleton. 

I think we just need to call the `ElementManager.loadElements();`before we run the tests in order to bootstrap the ElementManager with the necessary data. This way we won't need the `spyOn` 🔍 😄

Comment (grady-lad) [reply src/components/mass-calculator/MassCalculator.test.tsx]: Actually I had a look around and it seems that the `ElementManager` is used by multiple components. 

Maybe we should just initiate the `ElementManager` and load all the elements within `setupTests`. 

E.G.

``` javascript
// src/setupTests.ts

import ElementManager from "./ElementManager";
ElementManager.loadElements();
```

From what I can see `ElementManager` does not modify any of the data that it exposes. `ElementManager` only provides getter functions to access the data so I shouldn't be an issue call `ElementManager.loadElements();` within `setupTests`?

Comment (Aprillion) [reply src/components/mass-calculator/MassCalculator.test.tsx]: sounds like a good idea to `loadElements()` in setupTests.ts (unless we find some tests where we do NOT want to have it initialized of course)

if you do this, please don't forget update the test(s) that initialize it inside `beforeAll` - no need to do it twice.

Comment (Aprillion) [reply src/components/mass-calculator/MassCalculator.test.tsx]: @JacobMGEvans - I would vote against mocking `ElementManager`, IMHO the real thing works fine in tests, e.g. https://github.com/OpenSourceRaidGuild/atom-pwa/blob/next/src/components/element-picker/ElementPicker.test.tsx

File: src/components/mass-calculator/MassCalculator.test.tsx [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  expect(screen.getByText(/mass calculator/i)).toBeInTheDocument();
+  expect(screen.getByText(/total/i)).toBeInTheDocument();
+  expect(screen.getByText(/298\.06532 g \/ mol/i)).toBeInTheDocument();
+  expect(screen.getByText(/add element/i)).toBeInTheDocument();
+  expect(screen.getByText(/clear/i)).toBeInTheDocument();
+  expect(screen.getByText(/holmium/i)).toBeInTheDocument();
+  expect(screen.getByText(/164\.93032 g \/ mol/i)).toBeInTheDocument();
+  expect(screen.getByText(/ruthenium/i)).toBeInTheDocument();
+  expect(screen.getByText(/sulfur/i)).toBeInTheDocument();
+  expect(screen.getByText(/32\.065 g \/ mol/i)).toBeInTheDocument();
```
Comment (Aprillion) [inline src/components/mass-calculator/MassCalculator.test.tsx]: IMHO would be enough to assert only some subset of the above.

* then click on clear
* then click on add element and choose some element (e.g. oxygen)
* then add another element (e.g. hydrogen)
* then change some amount (e.g. hydrogen from 1 to 2)
* and assert on the expected total mass (e.g. 18.01528 g / mol for water as per https://en.wikipedia.org/wiki/Properties_of_water - and I checked the mass calculation for H2O manually and it outputs the correct mass, so I would expect this test to pass even if automated)

File: src/components/mass-calculator/MassCalculator.test.tsx (hunk 1 above)
Comment (mpeyper) [reply src/components/mass-calculator/MassCalculator.test.tsx]: Yeah, quite a few tests have been using the real version directly already without any issues. I suggest not mocking it and loading the elements up front is fine here.

Comment (MohitPopli) [reply src/components/mass-calculator/MassCalculator.test.tsx]: Thanks guys for suggestions. I'll add `ElementManager.loadElements()` inside setup file and will modify rest.

Comment (JacobMGEvans) [reply src/components/mass-calculator/MassCalculator.test.tsx]: Cool (Singleton! I knew I have seen that pattern before! lol) @grady-lad @Aprillion you can check out this comment too: https://github.com/OpenSourceRaidGuild/atom-pwa/pull/52#discussion_r584224172

File: src/components/mass-calculator/MassCalculator.test.tsx (hunk 2 above)
Comment (JacobMGEvans) [reply src/components/mass-calculator/MassCalculator.test.tsx]: Second this

**Review by mitchelvanbever — APPROVED** [review]

**Review by Aprillion — APPROVED** [review]
